dvc/cli.py

```diff
@@ -108,6 +145,13 @@ def error(self, message, cmd_cls=None):  # pylint: disable=arguments-differ
         _find_parser(self, cmd_cls)
 
     def parse_args(self, args=None, namespace=None):
+        if args is None:
+            args = sys.argv[1:]
+        else:
+            args = list(args)
+
+        self.find_cmd_suggestions(args)
+
```

Comment:
Do we need to do it before we parse the args though? Seems like we can do that after a failure (`if argv` down below).

Reply:
I thought so too but the reason I left `self.find_cmd_suggestions(args)` before the `parse_known_args` method was because `parse_known_args` throws its own internal error if the choices passed in are invalid. So the following:

```python
def parse_args(self, args=None, namespace=None):
    if args is None:
        args = sys.argv[1:]
    else:
        args = list(args)

    # NOTE: overriding to provide a more granular help message.
    # E.g. `dvc plots diff --bad-flag` would result in a `dvc plots diff`
    # help message instead of generic `dvc` usage.
    args, argv = self.parse_known_args(args, namespace)
    if argv:
        self._find_cmd_suggestions(args)

        msg = "unrecognized arguments: %s"
        self.error(msg % " ".join(argv), getattr(args, "func", None))
    return args
```

Leads to this:
<img width="1552" alt="Screen Shot 2020-10-24 at 6 04 49 PM" src="https://user-images.githubusercontent.com/7004949/97095319-96869180-1623-11eb-82ec-952161cd7490.png">

```python
2020-10-24 18:01:25,643 ERROR: argument cmd: invalid choice: 'dif' (choose from 'show', 'diff', 'modify')
```

[Python source](https://github.com/python/cpython/blob/b084d1b97e369293d2d2bc0791e2135822c923a8/Lib/argparse.py#L2499)

Reply:
@santos22 would it make sense to catch it and then process the suggestions?

Reply:
I think that would make sense but I'm not sure this is possible with Python 3.7.

The following still leads to the same error message:
```python
try:
    args, argv = self.parse_known_args(args, namespace)
except ArgumentError:
    self._find_cmd_suggestions(args)
else:
    if argv:
        msg = "unrecognized arguments: %s"
        self.error(msg % " ".join(argv), getattr(args, "func", None))
finally:
    return args
```

The `exit_on_error` argument when initializating `DvcParser` seems useful but it is available only in Python 3.9.

See https://docs.python.org/3/library/argparse.html#exit-on-error

Reply:
@efiop any further thoughts on this?

Reply:
@efiop thoughts on the current iteration?